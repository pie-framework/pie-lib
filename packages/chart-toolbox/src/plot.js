import { Root, Draggable, gridDraggable, utils, trig, types, createGraphProps } from './code/plot';

export { Root, Draggable, gridDraggable, utils, trig, types, createGraphProps };
