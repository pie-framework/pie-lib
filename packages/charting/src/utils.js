import { scaleBand } from '@vx/scale';
export const bandKey = (d, index) => `${index}-${d.label || '-'}`;

export const dataToXBand = (scaleX, data, width) => {
  return scaleBand({
    rangeRound: [0, width],
    domain: data.map(bandKey),
    padding: 0.2
  });
};
