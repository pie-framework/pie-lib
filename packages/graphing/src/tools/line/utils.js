import debug from 'debug';
// import { swap } from '../point/utils';

const log = debug('pie-lib:charting:line:utils');

const lineExpressionRegex = new RegExp(/([+|-]?\d+)?(x)?([+|-]\d+)?/i);

export class Expression {
  constructor(multiplier, b) {
    this.multiplier = multiplier;
    this.isVerticalLine = Math.abs(this.multiplier) === Infinity;
    this.b = this.isVerticalLine ? undefined : b;
  }

  expr() {
    return this.isVerticalLine ? 'x' : `${this.multiplier}x + ${this.b}`;
  }

  getY(x) {
    return this.isVerticalLine ? x : this.multiplier * x + this.b;
  }

  getX(y) {
    return this.isVerticalLine ? y : (y - this.b) / this.multiplier;
  }

  equals(other) {
    return this.multiplier === other.multiplier && this.b === other.b;
  }
}

export const pointsHaveSameExpression = (a, b) => {
  const expressionA = expression(a.from, a.to);
  const expressionB = expression(b.from, b.to);
  return expressionA.equals(expressionB);
};
/**
 * Create a linear expression from 2 points
 */

export const expression = (from, to) => {
  log('[expression] from: ', from, 'to: ', to);
  const multiplier = (to.y - from.y) / (to.x - from.x);
  const zeroedY = multiplier * from.x;
  const b = from.y - zeroedY;
  return new Expression(multiplier, b);
};

export const point = (x, y) => ({ x, y });

export const pointsFromExpression = expression => {
  const to = point(0, expression.b);
  const huh = expression.getX(0);
  const from = point(huh, 0);

  return {
    from,
    to
  };
};

export const expressionFromDescriptor = descriptor => {
  const lineDescriptor = descriptor.match(lineExpressionRegex);
  const maybeB = parseInt(lineDescriptor[3], 10);
  const maybeMultiplier = parseInt(lineDescriptor[1], 10);
  let multiplier = isNaN(maybeMultiplier)
    ? lineDescriptor[2] === undefined
      ? 0
      : 1
    : maybeMultiplier;
  let b = isNaN(maybeB) ? 0 : maybeB;

  // it's a constant, no variable found
  if (lineDescriptor[2] === undefined) {
    b = multiplier;
    multiplier = 0;
  }

  return new Expression(multiplier, b);
};

export const hasLine = (lines, line) => lineIndex(lines, line) !== -1;

export const removeLine = (lines, line) => {
  const index = lineIndex(lines, line);
  if (index === -1) {
    return lines;
  }
  const out = lines.slice();
  out.splice(index, 1);
  return out;
};

export const removeLines = (lines, toRemove) => {
  log('[removeLines] points: ', lines, toRemove);
  toRemove = Array.isArray(toRemove) ? toRemove : [toRemove];
  const out = lines.filter(p => !hasLine(toRemove, p));
  log('[removePoints] points: ', out);
  return out;
};

export const pointsEqual = (a, b) => a.x == b.x && a.y == b.y;

export const linesEqual = (a, b) => pointsEqual(a.from, b.from) && pointsEqual(a.to, b.to);

export const lineIndex = (lines, line) => lines.findIndex(l => linesEqual(l, line));

// export const swapLine = (lines, from, to) => swap(lines, from, to, lineIndex);
