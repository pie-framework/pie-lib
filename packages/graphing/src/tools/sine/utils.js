import _ from 'lodash';
import debug from 'debug';
import Point from '@mapbox/point-geometry';
const log = debug('pie-lib:graphing;sine:utils');

/**
 * build the x points with root as the anchor.
 * adds padding left and right so that
 * the anchor can be moved left/right
 * and there will be points off stage rendering.
 *
 *  root: 0, freq: 2, min: -5, max: 5 ==> -5,-4-2,0,2,4,5
 */
export const xPoints = (root, freq, min, max) => {
  return _.concat(
    min - freq,
    _.rangeRight(root, min - freq, freq * -1),
    _.range(root + freq, max + freq, freq),
    max + freq
  );
};

export const sinY = (amplitude, freq) => x => {
  const num = 2 * Math.PI * x;
  const frac = num / freq;
  return amplitude * parseFloat(Math.sin(frac).toFixed(10));
};

export const buildDataPoints = (min, max, root, interval, yFn) => {
  log('[buildDataPoints] min:', min, 'max:', max, 'root:', root);
  // const fn = sinY(amplitude, freq);
  const diff = max - root.x;
  const plusDiff = min - root.x;
  const xs = xPoints(0, interval, min - root.x - diff, max - root.x - plusDiff);
  log('xs:', xs);
  return xs.map(v => new Point(v, yFn(v))).map(p => p.add(new Point(root.x, root.y)));
};
