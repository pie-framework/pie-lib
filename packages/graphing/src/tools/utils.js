import debug from 'debug';
import Point from '@mapbox/point-geometry';
import _ from 'lodash';
const log = debug('pie-lib:graphing:tools:utils');
/**
 * build the x points with root as the anchor.
 */
export const xPoints = (root, freq, min, max) => {
  freq = Math.abs(freq);
  log('[xPoints] root:', root, 'freq:', freq, ' min:', min, ' max:', max);
  const leftpoints = _.rangeRight(root, min - freq, freq * -1);
  const rightpoints = _.range(root + freq, max + freq, freq);

  return _.concat(leftpoints, rightpoints);
};

/**
 * Calculate y for x value using amp, freq and shift
 * https://www.desmos.com/calculator/f1psfoiiv6
 */
export const sinY = (amplitude, freq, shift) => x => {
  shift = { phase: 0, vertical: 0, ...shift };
  const TWO_PI = Math.PI * 2;
  const num = TWO_PI * (x - shift.phase);
  const frac = num / freq;
  const sin = Math.sin(frac);
  const sinFloat = parseFloat(sin).toFixed(10);
  return amplitude * sinFloat + shift.vertical;
};

/**
 * You need three points for the quadratic eq, so we mirror the edge point
 * https://www.desmos.com/calculator/xhz1xqgpca
 */
export const parabolaFromTwoPoints = (root, edge) => {
  edge = edge || { ...root };
  const mirrorEdge = { x: root.x - (edge.x - root.x), y: edge.y };
  return parabolaFromThreePoints(root, edge, mirrorEdge);
};

export const parabolaFromThreePoints = (one, two, three) => {
  log(one, two, three);
  const { a, b, c } = pointsToABC(one, two, three);
  log(a, b, c);
  return parabola(a, b, c);
};
/** derive a,b,c from the 3 points */
export const pointsToABC = (one, two, three) => {
  const a1 = Math.pow(one.x, 2) * -1 + Math.pow(two.x, 2);
  const b1 = one.x * -1 + two.x;
  const d1 = one.y * -1 + two.y;
  const a2 = Math.pow(two.x, 2) * -1 + Math.pow(three.x, 2);
  const b2 = two.x * -1 + three.x;
  const d2 = two.y * -1 + three.y;
  const bMult = (b2 / b1) * -1;
  const a3 = bMult * a1 + a2;
  const d3 = bMult * d1 + d2;
  const a = d3 / a3;
  const b = (d1 - a1 * a) / b1;
  const c = one.y - a * Math.pow(one.x, 2) - b * one.x;
  return { a, b, c };
};

/**
 * y=ax^2+bx+c
 */
export const parabola = (a, b, c) => x => a * Math.pow(x, 2) + b * x + c;

/**
 * Generate a set of data points, add spacing before min and after max if there is space between minx + min and maxX and max
 * @param {*} min
 * @param {*} max
 * @param {*} root
 * @param {*} edge
 * @param {*} interval
 * @param {*} yFn
 */
export const buildDataPoints = (min, max, root, edge, interval, yFn) => {
  log('[buildDataPoints] min:', min, 'max:', max, 'root:', root);
  edge = edge ? edge : { ...root };
  const minX = Math.min(root.x, edge.x);
  const maxX = Math.max(root.x, edge.x);
  const leftSpace = min - minX;
  const rightSpace = max - maxX;
  console.log(minX, interval);
  const xs = xPoints(minX, interval, min - rightSpace, max - leftSpace);
  log('[buildDataPoints]:xs:', xs);
  return xs.map(v => new Point(v, yFn(v)));
};
