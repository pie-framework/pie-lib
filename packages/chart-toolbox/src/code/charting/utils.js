import { scaleBand, scalePoint } from '@vx/scale';
import { utils } from '../plot';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const bandKey = (d, index) => `${index}-${d.label || '-'}`;

export const dataToXBand = (scaleX, data, width, type) => {
  switch (type) {
    case 'bar':
    case 'dotPlot':
    case 'linePlot':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0.2,
      });

    case 'histogram':
      return scaleBand({
        rangeRound: [0, width],
        domain: data && data.map(bandKey),
        padding: 0,
      });

    case 'lineCross':
    case 'lineDot':
      return scalePoint({
        domain: data && data.map(bandKey),
        rangeRound: [0, width],
      });

    default:
      return scaleBand({
        range: [0, width],
        domain: data && data.map(bandKey),
        padding: 0,
      });
  }
};

export const getTickValues = (prop = {}) => {
  const tickValues = [];
  let tickVal = prop.min;

  while (tickVal <= prop.max) {
    tickValues.push(tickVal);
    tickVal = Math.round((tickVal + prop.step) * 10000) / 10000;
  }

  return tickValues;
};

export const getDomainAndRangeByChartType = (domain, range, chartType) => {
  let { step, labelStep, min, max } = range || {};

  if (!min) {
    min = 0;
  }

  if (!max || max < 0) {
    max = range.min + 1;
  }

  if (!step) {
    step = labelStep || 1;
  }
  if (!labelStep || (isNaN(labelStep) && step)) {
    labelStep = step || 1;
  }

  range.max = max;

  switch (chartType) {
    // if chart is dot plot or line plot, we should ignore step and make sure that min & max are integer values
    case 'dotPlot':
    case 'linePlot': {
      const intMin = Math.round(min);
      const intMax = Math.round(max);

      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1,
        },
        range: {
          ...range,
          min: intMin,
          max: intMin === intMax ? intMin + 1 : intMax,
          labelStep,
          step: 1,
        },
      };
    }

    default:
      return {
        domain: {
          ...domain,
          step: 1,
          labelStep: 1,
          min: 0,
          max: 1,
        },
        range: {
          ...range,
          labelStep,
          step,
        },
      };
  }
};

export const getGridLinesAndAxisByChartType = (range, chartType) => {
  switch (chartType) {
    case 'lineDot':
    case 'lineCross':
      return {
        verticalLines: undefined,
        horizontalLines: getTickValues(range),
        leftAxis: true,
      };

    case 'dotPlot':
    case 'linePlot':
      return {
        verticalLines: [],
        horizontalLines: [],
        leftAxis: false,
      };

    default:
      return {
        verticalLines: [],
        horizontalLines: getTickValues(range),
        leftAxis: true,
      };
  }
};

export const getRotateAngle = (fontSize, height) => {
  if (height >= fontSize * 2) {
    return 25;
  }

  return 0;
};

export const getTopPadding = (barWidth) => {
  if (barWidth < 30) {
    return 50;
  }

  if (barWidth < 40) {
    return 30;
  }

  if (barWidth < 60) {
    return 15;
  }

  return 0;
};

// This function calculates the transformation scale for SVG and the icon's vertical distance from its category
export const getScale = (width) => {
  let scale, deltay;

  if (width > 91) {
    scale = 1.3;
    deltay = -55;
  } else if (width > 45) {
    scale = 1.1;
    deltay = -45;
  } else if (width > 40) {
    scale = 0.5 + (width - 34) * 0.02;
    deltay = -25;
  } else if (width > 30) {
    scale = 0.5 + (width - 34) * 0.02;
    deltay = -20;
  } else {
    scale = 0.5 * Math.pow(0.98, 34 - width); // 0.98 is the reduction factor. Adjust to control scaling.
    deltay = -15;
  }

  return { scale, deltay };
};

export const getAdjustedX = (width, scaleValue) => {
  const innerWidthOriginal = 57;
  const effectiveInnerWidth = innerWidthOriginal * scaleValue;
  return (width - effectiveInnerWidth) / 2;
};
