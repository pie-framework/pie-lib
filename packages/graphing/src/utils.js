import head from 'lodash/head';
import tail from 'lodash/tail';
import { utils } from '@pie-lib/plot';
import invariant from 'invariant';
import isEqual from 'lodash/isEqual';

export const tickCount = utils.tickCount;
export const bounds = utils.bounds;
export const point = utils.point;

export const polygonToArea = points => {
  const h = head(points);
  const area = {
    left: h.x,
    top: h.y,
    bottom: h.y,
    right: h.x
  };
  return tail(points).reduce((a, p) => {
    a.left = Math.min(a.left, p.x);
    a.top = Math.max(a.top, p.y);
    a.bottom = Math.min(a.bottom, p.y);
    a.right = Math.max(a.right, p.x);
    return a;
  }, area);
};

export const lineToArea = (from, to) => pointsToArea(from, to);

export const pointsToArea = (a, b) => {
  invariant(!!a && !!b, 'a or b is undefined');
  const left = Math.min(a.x, b.x);
  const top = Math.max(a.y, b.y);
  const bottom = Math.min(a.y, b.y);
  const right = Math.max(a.x, b.x);
  return { left, top, bottom, right };
};

export const isDomainRangeEqual = (graphProps, nextGraphProps) => {
  return (
    isEqual(graphProps.domain, nextGraphProps.domain) &&
    isEqual(graphProps.range, nextGraphProps.range)
  );
};
