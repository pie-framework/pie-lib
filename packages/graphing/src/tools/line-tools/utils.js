import { calculateThirdPointOnLine } from '../../utils';

export const arrowDimensions = {
  none: 0,
  vector: 7,
  ray: 5,
  line: 5
};

export const calculateCorrectScaledPoints = (scale, from, to, type = 'none') => {
  const scaledFromX = scale.x(from.x);
  const scaledFromY = scale.y(from.y);
  let scaledToX = scale.x(to.x);
  let scaledToY = scale.y(to.y);
  let middlePoint;

  middlePoint = calculateThirdPointOnLine(
    { x: scaledToX, y: scaledToY },
    { x: scaledFromX, y: scaledFromY },
    {
      domain: {
        min: scale.x(to.x) - arrowDimensions[type],
        max: scale.x(to.x) + arrowDimensions[type]
      },
      range: {
        min: scale.y(to.y) - arrowDimensions[type],
        max: scale.y(to.y) + arrowDimensions[type]
      }
    }
  );

  if (middlePoint && isFinite(middlePoint.x) && isFinite(middlePoint.y)) {
    scaledToX = middlePoint.x;
    scaledToY = middlePoint.y;
  }

  return {
    x: scaledToX,
    y: scaledToY
  };
};
