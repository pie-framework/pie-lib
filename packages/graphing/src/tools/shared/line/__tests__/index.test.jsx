import { lineTool, lineToolComponent, lineBase } from '../index';
import { utils } from '@pie-lib/plot';
import { shallow } from 'enzyme';
import React from 'react';
import { graphProps as getGraphProps } from '../../../../__tests__/utils';
const { xy } = utils;
describe('lineTool', () => {
  describe('addPoint', () => {
    let toolbar;
    beforeEach(() => {
      toolbar = lineTool('lineType', () => <div />)();
    });
    it('returns a building mark', () => {
      const result = toolbar.addPoint(xy(1, 1));
      expect(result).toEqual({
        type: 'lineType',
        building: true,
        from: xy(1, 1)
      });
    });

    it('returns a complete mark', () => {
      const result = toolbar.addPoint(xy(1, 1), { from: xy(0, 0) });
      expect(result).toEqual({
        building: false,
        from: xy(0, 0),
        to: xy(1, 1)
      });
    });
  });
});

describe('lineToolComponent', () => {
  let Comp;
  let mark;
  let onChange;
  let w;
  const wrapper = extras => {
    const defaults = {
      mark,
      onChange: jest.fn(),
      graphProps: getGraphProps()
    };
    const props = { ...defaults, ...extras };
    return shallow(<Comp {...props} />);
  };

  beforeEach(() => {
    Comp = lineToolComponent(() => <text />);
    mark = { from: xy(0, 0), to: xy(1, 1) };
  });
  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });
  describe('logic', () => {
    describe('startDrag', () => {
      it('sets state', () => {
        w = wrapper();
        w.instance().startDrag();
        expect(w.state('mark')).toEqual(mark);
      });
    });

    describe('stopDrag/changeMark', () => {
      let update = { from: xy(2, 2), to: xy(4, 4) };
      beforeEach(() => {
        w = wrapper();
        w.instance().changeMark(update);
        w.instance().stopDrag();
      });
      it('calls onChange', () => {
        expect(w.instance().props.onChange).toHaveBeenCalledWith(mark, update);
      });
    });
  });
});

describe('lineBase', () => {
  let Comp;
  let w;
  let onChange;

  beforeEach(() => {
    Comp = lineBase(() => <text />);
  });

  const wrapper = extras => {
    const defaults = {
      onChange: jest.fn(),
      graphProps: getGraphProps(),
      from: xy(0, 0),
      to: xy(1, 1)
    };
    const props = { ...defaults, ...extras };
    return shallow(<Comp {...props} />);
  };
  describe('render', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    const assertCallsOnChange = (fn, args, expected) => {
      it('calls onChange', () => {
        const update = { from: xy(2, 2), to: xy(4, 4) };
        const w = wrapper();
        w.instance()[fn](...args);
        expect(w.instance().props.onChange).toHaveBeenCalledWith(expected);
      });
    };

    describe('dragComp', () => {
      const update = { from: xy(2, 2), to: xy(4, 4) };
      assertCallsOnChange('dragComp', [update], update);
    });

    describe('dragFrom', () => {
      assertCallsOnChange('dragFrom', [xy(2, 2)], { from: xy(2, 2), to: xy(1, 1) });
    });

    describe('dragTo', () => {
      assertCallsOnChange('dragTo', [xy(2, 2)], { from: xy(0, 0), to: xy(2, 2) });
    });
  });
});
