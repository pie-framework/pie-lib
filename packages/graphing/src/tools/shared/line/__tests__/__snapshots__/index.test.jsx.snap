// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lineBase render renders 1`] = `
<g>
  <GridDraggable
    from={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [MockFunction],
          "y": [MockFunction],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [MockFunction],
          "y": [MockFunction],
        },
      }
    }
    onDrag={[Function]}
    to={
      Object {
        "x": 1,
        "y": 1,
      }
    }
  />
  <WithStyles(GridDraggable)
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [MockFunction],
          "y": [MockFunction],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [MockFunction],
          "y": [MockFunction],
        },
      }
    }
    onDrag={[Function]}
    x={0}
    y={0}
  />
  <WithStyles(GridDraggable)
    angle={45}
    graphProps={
      Object {
        "domain": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "range": Object {
          "max": 1,
          "min": 0,
          "step": 1,
        },
        "scale": Object {
          "x": [MockFunction],
          "y": [MockFunction],
        },
        "size": Object {
          "height": 400,
          "width": 400,
        },
        "snap": Object {
          "x": [MockFunction],
          "y": [MockFunction],
        },
      }
    }
    onDrag={[Function]}
    x={1}
    y={1}
  />
</g>
`;

exports[`lineToolComponent snapshot renders 1`] = `
<Component
  from={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  graphProps={
    Object {
      "domain": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "range": Object {
        "max": 1,
        "min": 0,
        "step": 1,
      },
      "scale": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
      "size": Object {
        "height": 400,
        "width": 400,
      },
      "snap": Object {
        "x": [MockFunction],
        "y": [MockFunction],
      },
    }
  }
  onChange={[Function]}
  onDragStart={[Function]}
  onDragStop={[Function]}
  to={
    Object {
      "x": 1,
      "y": 1,
    }
  }
/>
`;
